// entradasya.2/backend/controllers/compraController.js
// controllers/compraController.js

const { getDB } = require('../config/db'); 
const Compra = require('../models/CompraModel');
const CompraDetalle = require('../models/CompraDetalleModel');

// Funci贸n auxiliar: Genera un token QR 煤nico
const generarQRToken = (id_compra, index) => {
    const baseToken = `ORD${id_compra}DET${index}`;
    const timestamp = Date.now().toString(36);
    const random = Math.random().toString(36).substring(2, 6);
    return `${baseToken}-${timestamp}${random}`.toUpperCase();
};

/**
 * Registra la Compra, el Pago y el Detalle en una 煤nica transacci贸n de BD.
 */
exports.registrarCompraCompleta = async (req, res) => {
    const { 
        id_usuario, 
        entradas_detalle, 
        pago,          
        total_compra,    
        notas = null
    } = req.body;
    
    let connection; 
    
    try {
        // 1. OBTENER EL POOL DE CONEXIONES Y LA CONEXIN
        const dbPool = getDB(); 
        connection = await dbPool.getConnection(); 

        await connection.beginTransaction(); 
        
        // --- PASO A: REGISTRAR EN LA TABLA 'compras' ---
        const compraData = {
            id_usuario: id_usuario,
            total: total_compra,
            estado: 'pendiente', 
            notas: notas
        };
        const nuevaCompra = await Compra.create(compraData, connection);
        const id_compra = nuevaCompra.id_compra;
        
        // --- PASO B: REGISTRAR EN LA TABLA 'pagos' ---
        const estadoPago = (pago.id_metodo === 3) ? 'confirmado' : 'iniciado'; 
        
        await connection.execute(
            'INSERT INTO pagos (id_compra, id_metodo, monto, estado, referencia_proveedor) VALUES (?, ?, ?, ?, ?)',
            [id_compra, pago.id_metodo, total_compra, estadoPago, pago.referencia_proveedor || null]
        );

        // --- PASO C: REGISTRAR EN LA TABLA 'compra_detalle' ---
        let primerQrToken = null; 
        
        const detallesParaInsertar = entradas_detalle.map((detalle, index) => {
            const qr_token = generarQRToken(id_compra, index + 1);
            if (index === 0) primerQrToken = qr_token;
            
            return {
                id_compra: id_compra,
                id_tipo: detalle.id_tipo, 
                //  CORRECCIN: Se fija la cantidad a 1, ya que cada objeto representa una entrada individual.
                cantidad: 1, 
                // Se eliminaron los '|| null' de las propiedades obligatorias
                precio_unit: detalle.precio_unit,
                subtotal: detalle.subtotal, 
                qr_token: qr_token
            };
        });
        
        // Aqu铆 estaba el error en el log, ahora se enviar谩 la 'cantidad' correcta.
        await CompraDetalle.bulkCreate(detallesParaInsertar, connection); 

        // 2. CONFIRMAR TRANSACCIN
        await connection.commit(); 
        
        // 3. Respuesta de 茅xito
        res.status(201).json({ 
            success: true, 
            message: 'Transacci贸n completada.',
            transactionId: id_compra,
            qrToken: primerQrToken 
        });

    } catch (error) {
        // 4. REVERTIR TRANSACCIN EN CASO DE ERROR
        if (connection) {
            await connection.rollback(); 
        }
        
        console.error('Error durante la transacci贸n de compra:', error);
        // Si el error es una violaci贸n de NOT NULL, puedes manejarlo espec铆ficamente
        let errorMessage = 'Error al registrar la compra. Los datos fueron revertidos.';
        if (error.code === 'ER_BAD_NULL_ERROR' || error.errno === 1048) {
             errorMessage = "Error de datos: una columna obligatoria no fue proporcionada (ej. id_tipo, precio_unit).";
        }
        
        res.status(500).json({ 
            success: false, 
            message: errorMessage, 
            error: error.message 
        });
    } finally {
        // 5. Liberar la conexi贸n
        if (connection) {
            connection.release();
        }
    }
};
